numpy/core/src/multiarray/einsum.c.src

```diff
@@ -2769,8 +2769,11 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
 
     /* Initialize the output to all zeros and reset the iterator */
     ret = NpyIter_GetOperandArray(iter)[nop];
-    Py_INCREF(ret);
     PyArray_AssignZero(ret, NULL);
```

Comment:
Can you convince me why this line should before the if and not after it?

Reply:
Doing  `PyArray_AssignZero` on `out` will fail  since `out`, with activated writeback semantics, will have its [`READONLY` flag set](http://www.numpy.org/devdocs/reference/c-api.array.html?highlight=pyarray_setwritebackifcopy). We must only write to the operand ndarray, the one whose `base` is `out` and has extra memory allocated for the writeback buffer. `ret` is subsequently reassigned to `out` since that is what must be returned after wrtieback resolution which resets the `READONLY` flag to its previous value.